frontend/src/components/MonacoEditor/useMonaco.ts

```diff
@@ -175,7 +175,24 @@ const useMonaco = async (lang: string) => {
     editorInstance: Editor.IStandaloneCodeEditor,
     content: string
   ) => {
-    if (editorInstance) editorInstance.setValue(content);
+    if (editorInstance) {
+      // use executeEdits API can preserve undo stack, allow user to undo/redo
+      editorInstance.executeEdits("delete-content", [
+        {
+          range: new monaco.Range(1, 1, 10000, 1),
```

Comment:
Not sure if "10000, 1" is an explicitly chosen number suggested by the doc or an arbitrary number. If it's the latter, worth adding a comment.

Reply:
add the comment below